File: Problems 1-9/problem4.py
def is_palindrome(n):
	if isinstance(n, str):
		# If we've been given a string, check if the reverse is the same
		return (n == n[::-1])
	else:
		# We're going to convert our input to a string and reverse it
		return (str(n) == str(n)[::-1])

def find_largest_prime_number(upper_limit = 1000):
	"""Find the largest palindrome made from the product of two 3-digit numbers"""
	max_palindrome = 0

	for x in xrange(0, upper_limit):
		for y in xrange(0, upper_limit):
			if (x * y) > max_palindrome and is_palindrome(x * y):
				max_palindrome = x * y
	
	print "The largest palindrome product of numbers under %d is %d\n", (upper_limit, max_palindrome)
	
	return max_palindrome

if __name__ == '__main__':
	find_largest_prime_number()

File: Problems 1-9/problem4.c
#include <stdio.h>

static int is_palindrome(unsigned int n);

int main()
{
	unsigned int i, j, max_palindrome = 0, upper_limit = 1000;

	for (i = 100; i < upper_limit; i++)
	{
		for (j = 100; j < upper_limit; j++)
		{
			unsigned int passing_val = i * j;

			if (is_palindrome(passing_val) 
				&& passing_val > max_palindrome)
			{
				max_palindrome = passing_val;
			}
		}
	}

	printf("The largest palindrome product of numbers under %u is %u\n", 
		upper_limit, max_palindrome);

	return 0;
}

int is_palindrome(unsigned int n)
{
	unsigned int reverse = 0, tester = n;

	while (tester)
	{
		//shift digits left and grab the rightmost digit
		reverse = 10 * reverse + (tester % 10);

		//shift digits right
		tester /= 10;
	}

	//Return a simple boolean
	return reverse == n;
}
